src/main/java/wolox/training/repositories/BookRepository.java

```diff
@@ -15,4 +17,12 @@
 
     Optional<Book> findByIsbn(String isbn);
 
+    @Query(
+        value =
+            "SELECT * FROM book WHERE (:genre IS NULL OR genre = CAST(:genre AS VARCHAR)) AND "
+                + "(:publisher IS NULL OR publisher = CAST(:publisher AS VARCHAR)) AND "
+                + "(:year IS NULL OR year = CAST(:year AS VARCHAR))",
+        nativeQuery = true)
+    List<Book> findByGenreAndPublisherAndYear(@Param("genre") String genre,
+        @Param("publisher") String publisher, @Param("year") String year);
```

Comment:
Why you are using a native query here? , you could use a jpa query. Try to avoid native queries unless you really need  them.

Reply:
I did it for learning purposes, because I read that sometimes JPQL is not enough to achieve desired results. If you wish, I can change that.